src/webgpu/api/validation/encoding/cmds/debug.spec.ts

```diff
@@ -0,0 +1,142 @@
+      size: { width: 16, height: 16, depth: 1 },
+      usage: GPUTextureUsage.OUTPUT_ATTACHMENT,
+    });
+  }
+
+  createEncoder(encoderType: EncoderType): Encoder {
+    assert(this.#commandEncoder === undefined);
+    switch (encoderType) {
+      case 'non-pass':
+        return this.device.createCommandEncoder({});
+      case 'render bundle':
+        return this.device.createRenderBundleEncoder({ colorFormats: ['rgba8unorm'] });
+      case 'compute pass':
+        this.#commandEncoder = this.device.createCommandEncoder({});
+        return this.#commandEncoder.beginComputePass({});
+      case 'render pass':
+        this.#commandEncoder = this.device.createCommandEncoder({});
+        return this.#commandEncoder.beginRenderPass({
+          colorAttachments: [
+            {
+              attachment: this.makeAttachmentTexture().createView(),
+              loadValue: { r: 1.0, g: 0.0, b: 0.0, a: 1.0 },
+            },
+          ],
+        });
+    }
+  }
+
+  finishEncoder(encoder: Encoder, encoderType: EncoderType) {
+    let commandBuffer: GPUCommandBuffer | undefined = undefined;
+    switch (encoderType) {
+      case 'non-pass': {
+        commandBuffer = (encoder as GPUCommandEncoder).finish({});
+        break;
+      }
+      case 'render bundle': {
+        const bundle = (encoder as GPURenderBundleEncoder).finish({});
+        const commandEncoder = this.device.createCommandEncoder({});
+        const pass = commandEncoder.beginRenderPass({
+          colorAttachments: [
+            {
+              attachment: this.makeAttachmentTexture().createView(),
+              loadValue: { r: 1.0, g: 0.0, b: 0.0, a: 1.0 },
+            },
+          ],
+        });
+        pass.executeBundles([bundle]);
+        pass.endPass();
+        commandBuffer = commandEncoder.finish({});
+        break;
+      }
+      case 'compute pass':
+      case 'render pass': {
+        assert(this.#commandEncoder !== undefined);
+        (encoder as GPUComputePassEncoder | GPURenderPassEncoder).endPass();
+        commandBuffer = this.#commandEncoder?.finish();
+        this.#commandEncoder = undefined;
+        break;
+      }
+    }
+    if (commandBuffer !== undefined) {
+      this.queue.submit([commandBuffer]);
+    }
+  }
+}
+
+export const g = makeTestGroup(F);
+
+g.test('debug_group_balanced')
+  .params(
+    params()
+      .combine(poptions('encoderType', kEncoderTypes))
+      .combine(poptions('pushCount', [0, 1, 2]))
+      .combine(poptions('popCount', [0, 1, 2]))
+  )
+  .fn(t => {
+    const encoder = t.createEncoder(t.params.encoderType);
+    const shouldError = t.params.popCount !== t.params.pushCount;
+    t.expectValidationError(() => {
+      for (let i = 0; i < t.params.pushCount; ++i) {
+        encoder.pushDebugGroup(`${i}`);
+      }
+      for (let i = 0; i < t.params.popCount; ++i) {
+        encoder.popDebugGroup();
+      }
+      t.finishEncoder(encoder, t.params.encoderType);
+    }, shouldError);
```

Comment:
nit: the expectValidationError should be just for the finishEncoder

Reply:
oops, I missed this, thanks for catching it